state/protocol/inmem/epoch.go

```diff
@@ -12,121 +12,34 @@ import (
-func (e Epoch) TargetDuration() (uint64, error) {
-	return e.enc.TargetDuration, nil
-}
-
-// TargetEndTime returns the desired real-world end time for this epoch, represented as
-// Unix Time (in units of seconds). This target is specified by the FlowEpoch smart contract in
-// the EpochSetup event and used by the Cruise Control system to moderate the block rate.
-func (e Epoch) TargetEndTime() (uint64, error) {
-	return e.enc.TargetEndTime, nil
-}
-
-func (e Epoch) Clustering() (flow.ClusterList, error) {
-	return e.enc.Clustering, nil
-}
-
-func (e Epoch) DKG() (protocol.DKG, error) {
-	if e.enc.DKG != nil {
-		return DKG{*e.enc.DKG}, nil
-	}
-	return nil, protocol.ErrNextEpochNotCommitted
-}
-
-func (e Epoch) Cluster(i uint) (protocol.Cluster, error) {
-	if e.enc.Clusters == nil {
-		return nil, protocol.ErrNextEpochNotCommitted
-	}
-
-	if i >= uint(len(e.enc.Clusters)) {
-		return nil, fmt.Errorf("no cluster with index %d: %w", i, protocol.ErrClusterNotFound)
-	}
-	return Cluster{e.enc.Clusters[i]}, nil
-}
-
-func (e Epoch) ClusterByChainID(chainID flow.ChainID) (protocol.Cluster, error) {
-	if e.enc.Clusters == nil {
-		return nil, protocol.ErrNextEpochNotCommitted
-	}
-
-	for _, cluster := range e.enc.Clusters {
-		if cluster.RootBlock.Header.ChainID == chainID {
-			return Cluster{cluster}, nil
-		}
-	}
-	chainIDs := make([]string, 0, len(e.enc.Clusters))
-	for _, cluster := range e.enc.Clusters {
-		chainIDs = append(chainIDs, string(cluster.RootBlock.Header.ChainID))
-	}
-	return nil, fmt.Errorf("no cluster with the given chain ID %v, available chainIDs %v: %w", chainID, chainIDs, protocol.ErrClusterNotFound)
-}
-
-func (e Epoch) FinalHeight() (uint64, error) {
-	if e.enc.FinalHeight != nil {
-		return *e.enc.FinalHeight, nil
-	}
-	return 0, protocol.ErrEpochTransitionNotFinalized
-}
-
-func (e Epoch) FirstHeight() (uint64, error) {
-	if e.enc.FirstHeight != nil {
-		return *e.enc.FirstHeight, nil
-	}
-	return 0, protocol.ErrEpochTransitionNotFinalized
-}
-
+// Epochs provides access to epoch data, backed by a rich protocol state entry.
 type Epochs struct {
-	enc EncodableEpochs
+	entry *flow.RichProtocolStateEntry
 }
 
 var _ protocol.EpochQuery = (*Epochs)(nil)
 
 func (eq Epochs) Previous() protocol.Epoch {
-	if eq.enc.Previous != nil {
-		return Epoch{*eq.enc.Previous}
+	if eq.entry.PreviousEpoch == nil {
+		return invalid.NewEpoch(protocol.ErrNoPreviousEpoch)
 	}
-	return invalid.NewEpoch(protocol.ErrNoPreviousEpoch)
+	return NewCommittedEpoch(eq.entry.PreviousEpochSetup, eq.entry.PreviousEpochCommit)
 }
+
 func (eq Epochs) Current() protocol.Epoch {
-	return Epoch{eq.enc.Current}
+	return NewCommittedEpoch(eq.entry.CurrentEpochSetup, eq.entry.CurrentEpochCommit)
 }
+
 func (eq Epochs) Next() protocol.Epoch {
-	if eq.enc.Next != nil {
-		return Epoch{*eq.enc.Next}
+	switch eq.entry.EpochPhase() {
+	case flow.EpochPhaseStaking:
+		return invalid.NewEpoch(protocol.ErrNextEpochNotSetup)
```

Comment:
Would suggest to document the sentinel `protocol.ErrNextEpochNotSetup` in the interface [`EpochQuery.Next`](https://github.com/onflow/flow-go/blob/e6ef40e3f0598544f5dfabcce20391d2a6e4a07d/state/protocol/epoch.go#L15-L17)